src/libAtomVM/memory.c

```diff
@@ -529,6 +569,34 @@ HOT_FUNC static term memory_shallow_copy_term(term t, term **new_heap, int move)
     }
 }
 
+void memory_heap_append_fragment(Heap *heap, HeapFragment *fragment, term mso_list)
+{
+    // The fragment we are appending may have next fragments
+    // So we take our current next and we add it to the tail of the passed list
+    if (heap->root->next) {
+        HeapFragment *tail = fragment;
+        while (tail->next != NULL) {
+            tail = tail->next;
+        }
+        tail->next = heap->root->next;
```

Comment:
I am not understanding why this isn't `tail->next = heap->root->next;`. there must be something I am missing.

Reply:
We could do that, but it's less effort to make passed fragment the first next fragment (after the root fragment, which is untouched). Indeed, if current root was made non-root and passed fragment was made the new root, this would require updating mso_list and heap_end pointer.